content/_data/filters.js

```diff
@@ -0,0 +1,60 @@
+'use strict';
+
+const path = require('path');
+const fs = require('fs');
+
+const doxray = require('doxray');
+
+// relative path from the root directory of remedocs
+// to the root directory of any package to be documented
+const remeDir = './';
+
+// relative path from remeDir to filters inside the project
+const filterDir = path.join(remeDir, 'src/filters/');
+
+// define the documentation regex…
+const doxOptions = {
+  regex: {
+    js: {
+      opening: /\/\*\s*@docs[^\n]*\n/m,
+      closing: /\*\//,
+      comment: /\/\*\s*@docs[^*]*\*+(?:[^/*][^*]*\*+)*\//gm,
+      ignore: /\/\*\s*ignore-@docs[\s\S]*/gm,
+    },
+  },
+};
+
+module.exports = () => {
+  const js = [];
+
+  // get the docs
+  fs.readdir(filterDir, (err, files) => {
+    if (err) {
+      throw err;
+    }
+
+    files
+      .filter((file) => file.endsWith('.js'))
+      .forEach((file) => {
+        const data = {};
+
+        const filePath = path.join(filterDir, file);
+        const docs = doxray([filePath], doxOptions);
+
+        data.name = file;
+        data.path = filePath;
+        data.slug = file.slice(0, file.indexOf('.'));
+        data.info = docs.patterns.find(
+          (pattern) => pattern.category === 'file',
+        );
+        data.title = data.info ? data.info.label : data.name;
+        data.patterns = docs.patterns.filter(
+          (pattern) => pattern.category !== 'file',
+        );
+
+        js.push(data);
+      });
+  });
+
+  return js;
```

Comment:
This can't possibly be right; the empty array `js` will be returned before the callback in `readdir` finishes, surely? Is this working for you, and if so, how?

Reply:
hmmm, it is working for me (and seemingly on the server), so I guess I need advice on what the fix would be? I really don't understand sync/async stuff yet - which I imagine is the category of issue here?

Reply:
So, the callback in `readdir` happens at an unspecified future point; I guess it's mutating the `js` array, which is being passed by reference, so as long as any code _using_ that array operates on it after that unspecified future point, it's OK, though kinda by accident. I think the Right Thing™ is to return a promise and resolve that promise when the callback finishes, and have anything operating on this array happen in a `then` (or `await`) off that promise.

Reply:
@wlonk are you willing to make that change? I'm not clean onn how promises work.

Reply:
Yes, I absolutely am, but I want to make sure that the calling context can handle that, too.